prov/hook/hook_ze/src/hook_ze.c

```diff
@@ -0,0 +1,1887 @@
+/*
+ * Copyright (c) 2021 Intel Corporation. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "ofi_prov.h"
+#include "ofi_iov.h"
+#include "ofi_atomic.h"
+#include "hook_prov.h"
+#include "hook_ze.h"
+
+static int hook_ze_iov_compare(struct ofi_rbmap *map, void *key, void *data)
+{
+	struct hook_ze_desc *desc = data;
+	struct iovec *iov = key;
+
+	if (ofi_iov_shifted_left(iov, &desc->iov))
+		return -1;
+	if (ofi_iov_shifted_right(iov, &desc->iov))
+		return 1;
+
+	return 0;
+}
+
+static int hook_ze_add_region(struct hook_ze_domain *domain,
+		const struct iovec *iov, struct hook_ze_desc **ze_desc)
+{
+	struct fi_mr_attr attr = {0};
+	struct iovec base_iov;
+	uint64_t iface, device, flags;
+	int ret = 0;
+
+	*ze_desc = ofi_buf_alloc(domain->mr_pool);
+	if (!*ze_desc)
+		return -FI_ENOMEM;
+
+	iface = ze_is_addr_valid(iov->iov_base, &device,
+				 &flags) ? FI_HMEM_ZE : FI_HMEM_SYSTEM;
+
+	if (iface != FI_HMEM_ZE) {
+		(*ze_desc)->desc = NULL;
+		(*ze_desc)->iov = *iov;
+		goto out;
+	}
+
+	ret = ze_hmem_get_base_addr(iov->iov_base, &base_iov.iov_base,
+				    &base_iov.iov_len);
```

Comment:
Unless I'm missing something, the only reason this provider is limited to ze support are the above 2 calls: ze_is_addr_valid and ze_hmem_get_base_addr.  These are both defined as generic ofi_hmem_ops.  It should be trivial to either configure the hook to use a specific index into ofi_hmem_ops, or to pick the index based on the current system options.

Unless there's a reason to restrict this to ze, please rename the hook from hook_ze to hook_xpu, and replace 'ze' with 'xpu' throughout.

Reply:
Yes, I agree. Ideally this would be generic. The question is how to select/pass through that index. Thoughts?

Reply:
We don't necessarily need to limit ourselves.  You could call ofi_get_hmem_iface().

Reply:
You could but the address lookup is expensive as it is and would have to iterate through all the different ifaces. Though if the system only has one interface on it, the other interfaces would be skipped so this doesn't really matter. However, if we're on a system that has CUDA and ZE memory on it but the application is only using ZE memory, every pointer would have to go through the CUDA lookup first before every querying it through ZE. The question would be is that worth it/a realistic use case to consider? If it's not realistic, then that's fine for me but that was the case I was trying to avoid by calling the generic function.

Reply:
Since the provider is for ze only (hook_ze is the name :)), it's sufficient to check for ze memory only. It's user error to use the hook for cuda based application.

Reply:
In real applications, I doubt you would find a system with mixed GPUs.  At best a test system might have them, but I doubt an end customer would.  The user could build libfabric with one of the GPUs disabled from the build.  We could also add runtime options to exclude a GPU.  But that should be hidden under the hmem_init call, so that it would apply to other parts of the code as well.

Reply:
we could add an option to the hook provider, either through the name "i.e. hook_xpu:ze, hook_zpu:cuda, ..." or some other mechanism.

Reply:
Ok, then if that's not a real use case and we're only looking at one GPU at a time, I'm fine with doing it that way. And if it's an issue later then it's future us' problem :P 
I'll have to either change the get_hmem_iface call or add a new one to retrieve the device + flags for the registration. Any preference? No one is calling ofi_get_hmem_iface as of right now I believe.

Reply:
or maybe hook_hmem to be more consistent with the terminology already in use.

Reply:
@jianxin Also a valid option. So many choices!